### ISSUE TITLE
[ER] std::mem::array_len

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust#80051]
This is an Enhancement Request, I'd like this since years. I'd like a const function (like `std::mem::size_of`) that given an array returns its length:

```rust
fn main() {
    use std::mem::array_len;
    let a = [0_u32; 10];
    let b = [0; array_len::<&a>()];
}
```

Note that in this code the array 'a' isn't a const, otherwise this suffices:

```rust
const fn array_len<T, const N: usize>(_: &[T; N]) -> usize { N }
```

The purporse of array_len is to avoid defining a third usize constant LEN when you only want two arrays to have the same length and you want DRY code.

In D language you don't need such special function because the normal `.length` attribute of arrays and vectors is usable in a constant context too:

```d
void main() {
    int[5] a; // An array with length fixed at compile-time, like a Rust array.
    int[a.length] b; // Another one.
}
```

But I think we can't have `array::len` const in Rust.

**Comment by lachlansneff** [comment]
https://doc.rust-lang.org/std/primitive.slice.html#method.len

**Comment by leonardo-m** [comment]
> https://doc.rust-lang.org/std/primitive.slice.html#method.len

Your suggestions doesn't work:

```rust
fn main() {
    let a = [0_u32; 10];
    let b = [0; a.len()];
}
```
Gives:

```rust
error[E0435]: attempt to use a non-constant value in a constant
 --> ...\test.rs:3:17
  |
3 |     let b = [0; a.len()];
  |                 ^ non-constant value

error: aborting due to previous error
```rust

**Comment by lachlansneff** [comment]
That's because `a` is not a const value. To use it in a const context, the value itself must be const.

```rust

fn main() {
    const A: [u32; 10] = [0_u32; 10];
    let b = [0; A.len()];
}

**Comment by leonardo-m** [comment]
Right, that's why I'm asking in this enhancement request for a new special function in the std lib. Special because I can't implement it in library code (I think). The array length is known at compile-time even if the array isn't const.

**Comment by scottmcm** [comment]
One version of this: `a.map(|_| 0)`.  That'll give you an array the same length as another one, consisting entirely of zeros.

---

I think this boils down to essentially asking for `typeof`, since you're basically asking for `typeof(a)::LEN`.  As such, I think the way this is doable with accepted features is _existential type_.  So it might, one day, be doable something like this:

```
trait ArrayLen { const LEN: usize; }
struct ArrayLenImpl<const N: usize>;
impl<const N: usize> ArrayLen for ArrayLenImpl<N> { const LEN: usize = N; }

type Magic = impl ArrayLen;
fn magic<T, const N: usize>(_: &[T; N]) -> Magic { ArrayLenImpl::<N> }

fn main() {
    let a = [0_u32; 10];
    magic(&a);
    let b = [0; Magic::LEN];
}
```

**Comment by camelid** [comment]
Note that `core::mem::size_of` is just a wrapper around a compiler-provided intrinsic – this would require a compiler change, not a library one.

**Comment by clarfonthey** [comment]
@scottmcm why is the conversion to `Magic` required there? Can't you just return `N`?

**Comment by scottmcm** [comment]
@clarfonthey Returning `N` would still be a runtime value, though, not a const-eval one, since `a` isn't a const-eval-time value.  Hence the indirection through _existential type_.

**Comment by clarfonthey** [comment]
@scottmcm Yeah, which is what `const fn` is for. Is there a reason why that couldn't be done that way?

**Comment by Kobata** [comment]
It seems like while the literal pattern here can't be expressed yet, it is now possible to essentially write out something equivalent, just indirectly:

```rust
const fn matching_array<T, U: Copy, const N: usize>(initializer: U, arr: &[T; N]) -> [U; N] {
  [initializer; N]
}
```
